packages/providers/src/provider.test.ts

```diff
@@ -156,6 +156,20 @@ describe('Provider', () => {
     expect(minGasPrice).toBeDefined();
   });
 
+  it('can change the provider url of the current instance', async () => {
+    const providerUrl1 = 'http://127.0.0.1:4000/graphql';
+    const providerUrl2 = 'http://127.0.0.1:8080/graphql';
+    const provider = new Provider(providerUrl1);
+    const spyGraphQLClient = jest.spyOn(GraphQL, 'GraphQLClient');
+
+    expect(provider.url).toBe(providerUrl1);
+    provider.connect(providerUrl2);
+    expect(provider.url).toBe(providerUrl2);
+    expect(spyGraphQLClient).toBeCalledWith(providerUrl2, undefined);
+
+    spyGraphQLClient.mockRestore();
```

Comment:
I tried explaining, but maybe it wasn't clear enough — the mock restoration _must_ have come before the first expect/assertion after the `connect` call because if somehow the assertion fails, the restoration will never occur.

```suggestion

    spyGraphQLClient.mockRestore();

    expect(provider.url).toBe(providerUrl2);
    expect(spyGraphQLClient).toBeCalledWith(providerUrl2, undefined);
```

Reply:
@arboleya oh I see! I missed that, my bad

Reply:
Maybe we just need a `try-catch` block here, because if I restore the mock before the assertion it will not work. The `mockRestore` call will 'reset' the function we are trying to inspect and hence the assertion will fail.

Reply:
The problem with using `afterEach` is that we don't want to mock this method for all the tests. We are only mocking it for one test. If we mock it in a `beforeEach`, we'll have to explicitly restore the mock in all the tests that _don't_ mock this method and need its actual implementation to be intact in order to work properly.

Reply:
You are right.

But assertions inside try-catch blocks can generate false positives unless we use control variables to ensure it doesn't. Another approach could be to have an `afterEach` hook, feel free to do as you prefer.

Control variables:

```ts
let error;

try {
  expect(provider.url).toBe(providerUrl2);
  expect(spyGraphQLClient).toBeCalledWith(providerUrl2);
} catch (err) {
  error = err;
}

spyGraphQLClient.mockRestore();

expect(error).toBeFalsy();
```

Using test hooks:

```ts
describe('Provider', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });
  // ...
})
```

Reply:
Figured that would be the way to go! Thanks :) This is a very elegant solution